_posts/2024-08-01-given-priority-change-3.5.md

```diff
@@ -0,0 +1,259 @@
+is to leverage the inheritance rules as outlined in point 8 of [the
+language
+reference](https://docs.scala-lang.org/scala3/reference/changed-features/implicit-resolution.html):
+
+```scala
+class General
+class Specific extends General
+
+class LowPriority:
+  given a:General()
+
+object NormalPriority extends LowPriority:
+  given b:Specific()
+
+def run =
+  import NormalPriority.given
+  val x = summon[General]
+  val _: Specific = x // <- b was picked
+```
+
+The idea is to enforce prioritization through the inheritance
+hierarchies of classes that provide `given` instances. By importing the
+`given` instances from the object with the highest priority, you can
+control which instance is selected by the compiler.
+
+#### Upcoming Bugfixes
+
+ Due to the release cut-off, Scala 3.5.0 may still report a
+warning that the above pattern will be ambiguous in Scala
+3.6. However, this warning is incorrect. The issue has already been
+resolved, and an upcoming bugfix release will improve the
+accuracy of warnings.
+
+
+### Outlook
+
+We are considering adding `-rewrite` rules that automatically insert
+explicit parameters when a change in choice is detected.
+
+
+## Towards Context Bounds as Givens
+
+We are gradually phasing out remaining uses of Scala 2 style
+`implicit`s in favor of the `given`/`using` syntax.  Scala 3.5 marks
+the first step in transitioning context bounds on type parameters to
+givens, with this transition expected to be completed in the upcoming
+Scala 3.6 release.
+
+Currently, context bounds on type parameters still desugar into
+`implicit` parameters:
+
+```scala
+def f[Element : Eq : Ordering] = summon[Eq[Element]].toOrdering 
+// expands to:
+def f[Element >: Nothing <: Any](implicit evidence$1: Eq[Element],
+      implicit evidence$2: Order[Element]): Ordering[Element] =
+      evidence$2.toOrdering
+```
+
+Prior to Scala 3.5, it was possible to pass `implicit` arguments
+explicitly for context bounds as if they were regular arguments. In
+Scala 3.5, however, these parameters must be qualified with `using`:
+
+```scala
+val eq: Eq[Int] = ???
+val ord: Order[Int] = ???
+f(eq, ord)       // ok in Scala < 3.5, error in 3.5
+f(using eq, ord) // ok in Scala 3.5
+```
+
+At this stage, the change does not affect the expansion of functions
+like `f` above, which still rely on `implicit` parameters. However,
+this is a crucial step towards facilitating the eventual transition to
+`given`s for context bounds in Scala 3.6.
+
+To assist with the migration to explicit `using` clauses, Scala 3.5
+provides an error message and offers automatic rewrites:
+
+```scala
+-- Error: bounds.scala:10:2 ----------------------------------------------
+10 |  f(eq, ord)  // error
+   |  ^
+   |Context bounds will map to context parameters.
+   |A `using` clause is needed to pass explicit arguments to them.
+   |This code can be rewritten automatically under -rewrite -source 3.4-migration.
```

Comment:
does it actually print `3.4-migration` in the error? that seems surprising

Reply:
Yes, and it'll do the right thing.